rfd/0080-tls-ping.md

```diff
@@ -0,0 +1,415 @@
+---
+authors: Gabriel Corado (gabriel.oliveira@goteleport.com)
+state: draft
+---
+
+# RFD 80 - TLS routing Ping
+
+## What
+
+Provide to TLS routing protocols a mechanism to prevent idle connections
+(without any data being sent or received) without interfering with the
+underlying protocol.
+
+## Why
+
+Idle connections are closed after a pre-defined period when deployed with load
+balancers. In some of those services (such as [AWS ELB](https://docs.aws.amazon.com/global-accelerator/latest/dg/introduction-how-it-works.html#about-idle-timeout) and [GlobalAccelerator](https://docs.aws.amazon.com/global-accelerator/latest/dg/introduction-how-it-works.html#about-idle-timeout)),
+having the TCP Keep-Alive configured is not enough to prevent the connections
+from being closed due to inactivity.
+
+Dropping the connections can directly impact user experience when they perform
+long-running commands such as database queries.
+
+**NOTE:** The connections are not dropped in those environments for protocols that
+already implement a mechanism like this (like [http2](https://httpwg.org/specs/rfc7540.html#PING)). However, not all protocols
+supported by Teleport provide such features.
+
+## Details
+
+To avoid having idle connections, we need to send at least one byte before the
+idle timeout period elapses. For protocols that that doesn't offer a mechanism
+for transmitting data in pre-defined intervals (avoiding the connection from
+becoming idle), it is necessary to wrap them into a different protocol.
+
+### Ping protocol
+
+A light protocol enables the connections to send ping packets filtered out on
+the receiver. These packets can be sent periodically, preventing the connection
+from becoming idle.
+
+The protocol packet is defined by the following:
+```
+| length | data           |
+| uint32 | `length` bytes |
+```
+
+The `length` field is an `uint32` encoded in network order (big-endian).
+
+#### Ping message
+
+The ping packet consists of a message where `length = 0`, and there is no `data`
+present. These messages are not visible to the reader since they have no
```

Comment:
They're still visible to the proxy, just not passed to the upstream, right?

Reply:
Yes, those messages will be "filtered" by the ping connection on the local proxy.